zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/listeners/task/TaskListenerTest.java

```diff
@@ -488,6 +489,221 @@ public void shouldProvideVariablesOfTaskCompletionFetchingOnlySpecifiedVariables
             job -> assertThat(job.getVariables()).containsExactly(Map.entry("foo", "overwritten")));
   }
 
+  @Test
+  public void shouldCompleteTaskWithTaskListenerWhenJobResultDeniedIsNotSet() {
+    // given
+    final long processInstanceKey =
+        createProcessInstance(createProcessWithCompleteTaskListeners(LISTENER_TYPE));
+
+    ENGINE.userTask().ofInstance(processInstanceKey).complete();
+    ENGINE
+        .job()
+        .ofInstance(processInstanceKey)
+        .withType(LISTENER_TYPE)
+        .withResult(new JobResult())
+        .complete();
+
+    // then
+    assertThat(
+            RecordingExporter.userTaskRecords()
+                .withProcessInstanceKey(processInstanceKey)
+                .limit(r -> r.getIntent() == UserTaskIntent.COMPLETED))
+        .extracting(Record::getIntent)
+        .describedAs(
+            "Ensure that `COMPLETE_TASK_LISTENER` events were triggered between user task"
+                + " `COMPLETING` and `COMPLETED` events")
+        .containsSubsequence(
+            UserTaskIntent.COMPLETING,
+            UserTaskIntent.COMPLETE_TASK_LISTENER,
+            UserTaskIntent.COMPLETED);
+  }
+
+  @Test
+  public void shouldCompleteUserTaskWhenTaskListenerAcceptsTheOperation() {
+    // given
+    final long processInstanceKey =
+        createProcessInstance(createProcessWithCompleteTaskListeners(LISTENER_TYPE));
+
+    ENGINE.userTask().ofInstance(processInstanceKey).complete();
+    ENGINE
+        .job()
+        .ofInstance(processInstanceKey)
+        .withType(LISTENER_TYPE)
+        .withResult(new JobResult().setDenied(false))
+        .complete();
+
+    // then
+    assertThat(
+            RecordingExporter.userTaskRecords()
+                .withProcessInstanceKey(processInstanceKey)
+                .limit(r -> r.getIntent() == UserTaskIntent.COMPLETED))
+        .extracting(Record::getIntent)
+        .describedAs(
+            "Ensure that `COMPLETE_TASK_LISTENER` events were triggered between user "
+                + "task `COMPLETING` and `COMPLETED` events")
+        .containsSubsequence(
+            UserTaskIntent.COMPLETING,
+            UserTaskIntent.COMPLETE_TASK_LISTENER,
+            UserTaskIntent.COMPLETED);
+  }
```

Comment:
🔧 These two tests verify the default behaviour where `result.denied = false`. We can merge them into a single test, with multiple jobs completed using different `result` arguments. This could look something like:
<details>
  <summary>Test setup</summary>

```java
    final long processInstanceKey =
        createProcessInstance(
            createProcessWithCompleteTaskListeners(
                LISTENER_TYPE, LISTENER_TYPE + "_2", LISTENER_TYPE + "_3"));
    // when
    ENGINE.userTask().ofInstance(processInstanceKey).complete();

    ENGINE.job().ofInstance(processInstanceKey).withType(LISTENER_TYPE).complete();
    ENGINE
        .job()
        .ofInstance(processInstanceKey)
        .withType(LISTENER_TYPE + "_2")
        .withResult(new JobResult())
        .complete();

    ENGINE
        .job()
        .ofInstance(processInstanceKey)
        .withType(LISTENER_TYPE + "_3")
        .withResult(new JobResult().setDenied(false))
        .complete();
```
</details>

Then we could assert that all completed jobs have `denied = false`, like this:

<details>
  <summary>assertion</summary>

```java
    assertThat(
            RecordingExporter.jobRecords()
                .withProcessInstanceKey(processInstanceKey)
                .withJobKind(JobKind.TASK_LISTENER)
                .withJobListenerEventType(JobListenerEventType.COMPLETE)
                .withIntent(JobIntent.COMPLETED)
                .limit(3))
        .extracting(Record::getValue)
        .extracting(JobRecordValue::getType, v -> ((JobRecord) v).getResult().isDenied())
        .describedAs("Verify that all task listeners were completed with `denied=false`")
        .containsExactly(
            tuple(LISTENER_TYPE, false),
            tuple(LISTENER_TYPE + "_2", false),
            tuple(LISTENER_TYPE + "_3", false));
```
</details>

---
💭 Since this test covers the same behavior with different arguments, it would be ideal to use parameterized testing, but we’re limited to JUnit 4 here :smiling_face_with_tear:

Reply:
👍 Great suggestion, thank you @ce-dmelnych! And yes, parameterized test would be great